pjsip/src/pjsip/sip_auth_client.c

```diff
@@ -162,158 +196,107 @@ static void digestNtoStr(const unsigned char digest[], int n, char *output)
-                                            const pj_str_t *realm,
-                                            const pjsip_cred_info *cred_info,
-                                            const pj_str_t *method)
-{
-#if PJSIP_AUTH_HAS_DIGEST_SHA256
+    digest_len = algorithm->digest_length;
+    digest_strlen = algorithm->digest_str_length;
+    dig_len = digest_len;
 
-    char ha1[PJSIP_SHA256STRLEN];
-    char ha2[PJSIP_SHA256STRLEN];
-    unsigned char digest[32];
+    if (result->slen < digest_strlen) {
+        PJ_LOG(4, (THIS_FILE,
+                "The length of the result buffer must be at least %d bytes "
+                "for algorithm %.*s", digest_strlen,
+                (int)algorithm->iana_name.slen, algorithm->iana_name.ptr));
+        return PJ_EINVAL;
+    }
+    result->slen = 0;
 
-#if OPENSSL_VERSION_NUMBER < 0x30000000L
-    SHA256_CTX pms;
-#else
-    EVP_MD_CTX* mdctx;
-    const EVP_MD* md;
-    unsigned dig_len;
-#endif
+    if (!PJSIP_CRED_DATA_IS_PASSWD(cred_info) && !PJSIP_CRED_DATA_IS_DIGEST(cred_info)) {
+        PJ_LOG(4, (THIS_FILE,
+                "cred_info->data_type must be PJSIP_CRED_DATA_PLAIN_PASSWD "
+                "or PJSIP_CRED_DATA_DIGEST"));
+        return PJ_EINVAL;
+    }
 
-    pj_assert(result->slen >= PJSIP_SHA256STRLEN);
+    if (PJSIP_CRED_DATA_IS_DIGEST(cred_info)) {
+        if (cred_info->algorithm_type != algorithm_type) {
+            PJ_LOG(4,(THIS_FILE,
+                    "The algorithm specified in the cred_info (%.*s) "
+                    "doesn't match the algorithm requested for hashing (%.*s)",
+                    (int)pjsip_auth_algorithms[cred_info->algorithm_type].iana_name.slen,
+                    pjsip_auth_algorithms[cred_info->algorithm_type].iana_name.ptr,
+                    (int)pjsip_auth_algorithms[algorithm_type].iana_name.slen,
+                    pjsip_auth_algorithms[algorithm_type].iana_name.ptr));
+            return PJ_EINVAL;
+        }
+        PJ_ASSERT_RETURN(cred_info->data.slen >= digest_strlen, PJ_EINVAL);
+    }
 
-#if OPENSSL_VERSION_NUMBER >= 0x30000000L
-    md = EVP_get_digestbyname("SHA256");
+    md = EVP_get_digestbyname(algorithm->openssl_name);
     if (md == NULL) {
+        /* Shouldn't happen since it was checked above */
         return PJ_ENOTSUP;
     }
-#endif
 
-    AUTH_TRACE_((THIS_FILE, "Begin creating digest"));
+    AUTH_TRACE_((THIS_FILE, "Begin creating %.*s digest",
+            (int)algorithm->iana_name.slen, algorithm->iana_name.ptr));
 
-    if ((cred_info->data_type & PASSWD_MASK) == PJSIP_CRED_DATA_PLAIN_PASSWD)
+    if (PJSIP_CRED_DATA_IS_PASSWD(cred_info))
     {
+        AUTH_TRACE_((THIS_FILE, " Using plain text password for %.*s digest",
+                (int)algorithm->iana_name.slen, algorithm->iana_name.ptr));
         /***
-         *** ha1 = SHA256(username ":" realm ":" password)
+         *** ha1 = (digest)(username ":" realm ":" password)
          ***/
-#if OPENSSL_VERSION_NUMBER < 0x30000000L
```

Comment:
Looks like all codes for OpenSSL < 3 is removed, we expect it is still compatible with OpenSSL 1.1.x, can you confirm or have you tested?

Reply:
The code was a bit weird when it came to OpenSSL versions.  The things that were only enabled with OpenSSL >= 3.0.0 (EVP_MD, EVP_Digest, etc) were all available with earlier versions back to 1.1.0 so I'm not sure why those SHA256_ alternatives were there.

I did build Asterisk with this patch on CentOS7 with OpenSSL 1.1.0 and tested both UAC and UAS authentication, including detecting that OpenSSL 1.1.0 didn't support SHA-512-256.  Everything worked fine.

Reply:
From a manual page [here](https://docs.openssl.org/1.0.2/man3/EVP_DigestInit/#history), `EVP_DigestInit_ex()` seems to be introduced in 0.9.7. So perhaps those alternatives were actually for OpenSSL older than 0.9.7, which is too old by now, especially for a security library :)

Also thanks for the confirmation that it works fine with 1.1.0.